v10/cmd/lcc/etc/lcc.1
.TH LCC 1 "local \- 4/13/93"
.SH NAME
lcc \- local ANSI C compiler
.SH SYNOPSIS
.B lcc
[
.I option
|
.I file
]...
.br
.SH DESCRIPTION
.PP
.I lcc
is a local ANSI C compiler for the VAX, MIPS, SPARC, and Motorola MC68020.
.PP
Arguments whose names end with `.c' are taken to be
C source programs; they are preprocessed, compiled, and
each object program is left on the file
whose name is that of the source with `.o' substituted for `.c'.
Arguments whose names end with `.i' are treated similarly,
except they are not preprocessed.
In the same way,
arguments whose names end with `.s' are taken to be assembly source programs
and are assembled, producing a `.o' file.
.PP
.I lcc
deletes a `.o' file if and only if exactly one
source file (`.c', `.s', or `.i' file) is mentioned and no other file
(source, object, library) or
.B \-l
option is mentioned.
.PP
.I lcc
uses ANSI standard header files
in preference to the `old-style' header files normally found in
.B /usr/include
(see `FILES' below).
Include files not found in the ANSI header files
are taken from the normal default include areas,
which usually includes
.BR /usr/include .
.PP
.I lcc
interprets the following options; unrecognized options are
taken as loader options (see
.IR ld (1))
unless
.BR \-c ,
.BR \-S ,
or
.B \-E
precedes them.
Except for
.BR \-l ,
all options are processed before any of the files
and apply to all of the files.
Applicable options are passed to each compilation phase in the order given.
.TP
.B \-c
Suppress the loading phase of the compilation, and force
an object file to be produced even if only one program is compiled.
.TP
.B \-g
Produce additional symbol table information for the local debuggers.
.I lcc
warns when
.B \-g
is unsupported.
.TP
.B \-w
Suppress warning diagnostics, such as those
announcing unreferenced statics, locals, and parameters.
The line
.I
#pragma ref id
simulates a reference to the variable 
.IR id .
.TP
.BI \-d n
Generate jump tables for switches whose density is at least
.IR n ,
a floating point constant between zero and one.
The default is 0.5.
.TP
.SM
.B \-A
Warns about
declarations and casts of function types without prototypes,
missing return values in returns from int functions,
assignments between pointers to ints and pointers to enums, and
conversions from pointers to smaller integral types.
A second
.SM
.B \-A
warns about
unrecognized control lines,
non-ANSI language extensions and source characters in literals,
unreferenced variables and static functions,
declaring arrays of incomplete types,
and exceeding
.I some
ANSI environmental limits, like more than 257 cases in switches.
It also arranges for duplicate global definitions in separately compiled
files to cause loader errors.
.TP
.SM
.B \-P
Writes declarations for all defined globals on standard error.
Function declarations include prototypes;
editing this output can simplify conversion to ANSI C.
This output may not correspond to the input when
there are several typedef's for the same type.
.TP
.B \-n
Arrange for the compiler to produce code
that tests for dereferencing zero pointers.
The code reports the offending file and line number and calls
.IR abort (3).
.TP
.SM
.B \-O
is ignored.
.TP
.SM
.B \-S
Compile the named C programs, and leave the
assembler-language output on corresponding files suffixed `.s'.
.TP
.SM
.B \-E
Run only the preprocessor on the named C programs
and unsuffixed file arguments,
and send the result to the standard output.
.TP
.SM
.B \-C
Prevent the preprocessor from eliding comments.
.TP
.BI \-o "  output"
Name the output file
.IR output .
If
.B \-c
or
.B \-S
is specified and there is exactly one source file,
this option names the object or assembly file, respectively.
Otherwise, this option names the final executable
file generated by the loader, and `a.out' is left undisturbed.
.I lcc
warns if
.B \-o
and
.B \-c
or
.B \-S
are given with more than one source file and ignores the
.B \-o
option.
.TP
.SM
.BI \-D \*Sname=def
.br
.ns
.TP
.SM
.BI \-D \*Sname
Define the
.I name
to the preprocessor, as if by `#define'.
If no definition is given, the name is defined as "1".
.TP
.SM
.BI \-U \*Sname
Remove any initial definition of
.IR name .
.TP
.SM
.BI \-I \*Sdir
`#include' files
whose names do not begin with `/' are always
sought first in the directory of the
.I file
arguments, then in directories named in
.SM
.B \-I
options, then in directories on a standard list.
.TP
.SM
.B \-N
Do not search
.I any
of the standard directories for `#include' files.
Only those directories specified by explicit
.SM
.B \-I
options will be searched, in the order given.
.TP
.SM
.BI \-B \*Sstr
Use the compiler \fIstr\fP\fBrcc\fP instead of the default version.
Note that
.I str
often requires a trailing slash.
On Suns only,
.SM
.B \-Bstatic
and
.SM
.BI \-Bdynamic
are passed to the loader; see
.IR ld (1).
.TP
.B \-v
Print commands as they are executed; some of the executed
programs are directed to print their version numbers.
More than one occurrence of
.B \-v
causes the commands to be printed, but
.I not
executed.
.TP
.B \-b
Produce code that counts the number of times each expression is executed.
If loading takes place, replace the standard exit
function by one that writes a
.B prof.out
file when the object program terminates.
A listing annotated with execution counts can then be generated with
.IR bprint (1).
.I lcc
warns when
.B \-b
is unsupported.
.B \-Wf-C
is similar, but counts only the number of function calls.
.TP
.B \-p
Produce code that counts the number of times each function is called.
If loading takes place, replace the standard startup
function by one that automatically calls
.IR monitor (3)
at the start and arranges to write a
.B mon.out
file when the object program terminates normally.
An execution profile can then be generated with
.IR prof (1).
.I lcc
warns when
.B \-p
is unsupported.
.TP
.B \-pg
Causes the compiler to produce counting code like
.BR \-p ,
but invokes a run-time recording mechanism that keeps more
extensive statistics and produces a 
.B gmon.out
file at normal termination.
Also, a profiling library is searched, in lieu of the standard C library.
An execution profile can then be generated with
.IR gprof (1).
.I lcc
warns when
.B \-pg
is unsupported.
.TP
.SM
.BI \-t \*Sname
.br
.ns
.TP
.SM
.BI \-t
Produce code to print the name of the function, an activation number,
and the name and value of each argument at function entry.
At function exit, produce code to print
the name of the function, the activation number, and the return value.
By default,
.I printf
does the printing; if
.I name
appears, it does.
For null
.I char*
values, "(null)" is printed. 
On Suns only,
.SM
.BI \-target
.I name
is accepted, but ignored.
.TP
.SM
.BI \-W \*Sx \fIarg\fP
pass argument
.I arg
to the program indicated by
.IR x ;
.I x
can be one of
.BR p ,
.BR f ,
.BR a ,
or
.BR l ,
which refer, respectively, to the preprocessor, the compiler proper,
the assembler, and the loader.
.I arg
is passed as given; if a
.B \-
is expected, it must be given explicitly.
.SM
.BI \-W \*So \fIarg\fP
specifies a system-specific option,
.IR arg .
.PP
Other arguments
are taken to be either loader option arguments, or C-compatible
object programs, typically produced by an earlier
.I lcc
run, or perhaps libraries of C-compatible routines.
Duplicate `.o' files are ignored.
These programs, together with the results of any
compilations specified, are loaded (in the order
given) to produce an executable program with name
.BR a.out .
.PP
.I lcc
assigns the most frequently referenced scalar parameters and
locals to registers whenever possible.
For each block,
explicit register declarations are obeyed first;
remaining registers are assigned to automatic locals if they
are `referenced' at least 3 times.
Each top-level occurrence of an identifier
counts as 1 reference. Occurrences in a loop,
either of the then/else arms of an if statement, or a case
in a switch statement each count, respectively, as 10, 1/2, or 1/10 references.
These values are increased accordingly for nested control structures.
.B \-Wf-a
causes
.I lcc
to read a
.B prof.out
file from a previous execution and to use the data therein
to compute reference counts (see
.BR \-b ).
.SH LIMITATIONS
.PP
.I lcc
accepts the C programming language
as described in the proposed ANSI standard
and in the second edition of Kernighan and Ritchie.
.I lcc
is intended to be used with the GNU C preprocessor, which supports the
preprocessing features introduced by the ANSI standard.
The
.SM
.B \-Wp-trigraphs
option is required to enable trigraph sequences.
.PP
Wide-character literals are accepted
but are treated as plain char literals.
Plain chars are signed chars,
ints and long ints are the same size
as are doubles and long doubles, and
plain int bit fields are signed.
Bit fields are aligned like unsigned integers but are otherwise laid out
as if by the standard C compiler,
.IR cc (1).
Other compilers, such as the GNU C compiler,
.IR gcc (1),
may choose other, incompatible layouts.
.PP
Likewise, calling conventions are intended to be compatible with
.IR cc (1),
except possibly for passing and returning structures.
Specifically,
.I lcc
passes structures like
.I cc
on all targets,
but returns structures like
.I cc
on only the MIPS.
Consequently, calls to/from such functions compiled with
.I cc
or other C compilers may not work.
Calling a function that returns
a structure without declaring it as such violates
the ANSI standard and may cause a core dump.
.SH FILES
.PP
The file names listed below are
.IR typical ,
but vary among installations; installation-dependent variants
can be displayed by running
.I lcc
with the
.B \-v
option.
.PP
.ta \w'/usr/lib/gcc-cppxx'u
.nf
file.c	input file
file.o	object file
a.out	loaded output
/tmp/lcc?	temporaries
/usr/lib/gcc-cpp	preprocessor
/usr/lib/rcc	compiler
/usr/lib/bbexit.o	exit for profiling
/lib/crt0.o	runtime startoff
/lib/[gm]crt0.o	startoffs for profiling
/lib/libc.a	standard library
/usr/include/lcc	ANSI standard headers
/usr/include/libc	local ANSI headers
/usr/include	traditional headers
prof.out	file produced for \fIbprint\fR(1)
mon.out	file produced for \fIprof\fR(1)
gmon.out	file produced for \fIgprof\fR(1)
.fi
.PP
.I lcc
predefines the macros `unix' and `__LCC__'.
It may also predefine some installation-dependent symbols; option
.B \-v
exposes them.
.SH "SEE ALSO"
B. W. Kernighan and D. M. Ritchie,
.I The C Programming Language,
Prentice-Hall, 2nd Ed., 1988.
.PP
.I
American National Standard for Information Systems, Programming Language C,
ANSI X3.159-1989, American National Standard Institute, Inc., New York, 1990.
.PP
cc(1), ld(1)
.br
.SH BUGS
Mail bug reports along with the shortest program
that exposes them and the details reported by
.IR lcc 's
.B \-v
option to lcc-bugs@princeton.edu.
.PP
The `ANSI standard headers' conform to the specifications in
the standard, which may be too restrictive for some applications,
but necessary for portability.
Functions given in the ANSI headers may be missing from
the local C library (e.g., `wide character' functions)
or may not correspond exactly to the local version;
for example, the ANSI standard
.B stdio.h
specifies that
.IR printf ,
.IR fprintf ,
and
.I sprintf
return the number of characters written to the file or array,
but few existing libraries implement this convention.
.PP
On VAXes running 4.3bsd UNIX,
the definitions of
.I tolower
and
.I toupper
erroneously return bad values for non-letter arguments.
.PP
On the MIPS and SPARC, old-style variadic functions must use
.B varargs.h
from MIPS or Sun. New-style is recommended.
.PP
The Sun assembler version 3.5 or later is required on the MC68020,
and only the MC68881 is supported.
.PP
With
.BR \-b ,
files compiled
.I without
.B \-b
may cause
.I bprint
to print erroneous call graphs.
For example, if
.B f
calls
.B g
calls
.B h
and
.B f
and
.B h
are compiled with
.BR \-b ,
but
.B g
is not,
.B bprint
will report that
.B f
called
.BR h .
The total number of calls is correct, however.


v10/cmd/lcc/etc/lcc.c
/*
 * lcc [ option ]... [ file | -llib ]...
 * front end for the ANSI C compiler
 */
/* Fri Feb 5 07:56:38 EST 1993 */
#define VERSION ((2<<8)|0)

#include <stdio.h>
#include <ctype.h>
#include <signal.h>

#ifndef TEMPDIR
#define TEMPDIR "/tmp"
#endif

typedef struct list *List;
struct list {		/* circular list nodes: */
	char *str;		/* option or file name */
	List link;		/* next list element */
};

#if __LCC__ || __STDC__
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#define dclproto(func,args) func args
dclproto(static void interrupt,(int));
#else
#define assert(e) ((void)((e)||(fprintf(stderr, "assertion failed: file %s, line %d\n", \
	__FILE__, __LINE__), abort(), 0)))
#define dclproto(func,args) func()
dclproto(static void interrupt,());
#endif

dclproto(static char *alloc,(int));
dclproto(static List append,(char *,List));
dclproto(extern char *basename,(char *));
dclproto(static int callsys,(char *[]));
dclproto(extern char *concat,(char *, char *));
dclproto(static int compile,(char *, char *));
dclproto(static void compose,(char *[], List, List, List));
dclproto(static void cprint,(char *[], char *));
dclproto(static void error,(char *, char *));
dclproto(static void execute,(char *[]));
dclproto(static int exists,(char *));
dclproto(static int filename,(char *, char *));
dclproto(static List find,(char *, List));
dclproto(static void opt,(char *));
dclproto(static void rm,(List));
dclproto(extern char *strsave,(char *));
dclproto(extern int suffix,(char *));

dclproto(extern int execv,(char *, char *[]));
dclproto(extern int fork,(void));
dclproto(extern int close,(int));
dclproto(extern int dup,(int));
dclproto(extern int fork,(void));
dclproto(extern int getpid,(void));
dclproto(extern int open,(char *, int));
dclproto(extern int pipe,(int[]));
dclproto(extern int read,(int, char *, int));
dclproto(extern int unlink,(char *));
dclproto(extern int wait,(int*));

extern char *cpp[], *include[], *com[], *as[],*ld[];
dclproto(extern int option,(char *));

static int errcnt;		/* number of errors */
static int Eflag;		/* -E specified */
static int Sflag;		/* -S specified */
static int cflag;		/* -c specified */
static int verbose;		/* incremented for each -v */
static List llist[2];		/* loader files, flags */
static List alist;		/* assembler flags */
static List clist;		/* compiler flags */
static List plist;		/* preprocessor flags */
static List rmlist;		/* list of files to remove */
static char *outfile;		/* ld output file or -[cS] object file */
static int ac;			/* argument count */
static char **av;		/* argument vector */
static char tempname[sizeof TEMPDIR + 15];	/* temporary .s file */
static char *progname;

main(argc, argv) char *argv[]; {
	int i, j, nf;
	
	progname = argv[0];
	ac = argc + 50;
	av = (char **)alloc(ac*sizeof(char *));
	if (signal(SIGINT, SIG_IGN) != SIG_IGN)
		signal(SIGINT, interrupt);
	if (signal(SIGTERM, SIG_IGN) != SIG_IGN)
		signal(SIGTERM, interrupt);
#ifdef SIGHUP
	if (signal(SIGHUP, SIG_IGN) != SIG_IGN)
		signal(SIGHUP, interrupt);
#endif
	sprintf(tempname, "%s/lcc%05d.s", TEMPDIR, getpid());
	rmlist = append(tempname, rmlist);
	plist = append("-D__LCC__", append("-Dunix", 0));
	for (nf = 0, i = j = 1; i < argc; i++) {
		if (strcmp(argv[i], "-o") == 0) {
			if (++i < argc) {
				if (strchr("ci", suffix(argv[i]))) {
					error("-o would overwrite %s", argv[i]);
					exit(8);
				}
				outfile = argv[i];
				continue;
			} else {
				error("unrecognized option `%s'", argv[i]);
				exit(8);
			}
#ifdef sun
		} else if (strcmp(argv[i], "-target") == 0) {
			if (argv[i+1] && *argv[i+1] != '-')
				i++;
			continue;
#endif
		} else if (*argv[i] == '-' && argv[i][1] != 'l') {
			opt(argv[i]);
			continue;
		} else if (*argv[i] != '-' && strchr("csi", suffix(argv[i])))
			nf++;
		argv[j++] = argv[i];
	}
	if ((cflag || Sflag) && outfile && nf != 1) {
		fprintf(stderr, "%s: -o %s ignored\n", progname, outfile);
		outfile = 0;
	}
	argv[j] = 0;
	for (i = 0; include[i]; i++)
		plist = append(include[i], plist);
	for (i = 1; argv[i]; i++)
		if (*argv[i] == '-')
			opt(argv[i]);
		else {
			if (nf > 1 && strchr("csi", suffix(argv[i]))) {
				fprintf(stderr, "%s:\n", argv[i]);
				fflush(stdout);
			}
			filename(argv[i], 0);
		}
	if (errcnt == 0 && !Eflag && !Sflag && !cflag && llist[1]) {
		if (i == 2 && strchr("csi", suffix(argv[1])))
			rmlist = append(concat(basename(argv[1]), ".o"), rmlist);
		compose(ld, llist[0], llist[1], append(outfile ? outfile : "a.out", 0));
		if (callsys(av))
			errcnt++;
	}
	rm(rmlist);	
	exit(errcnt > 0);
}

/* alloc - allocate n bytes or die */
static char *alloc(n) {
	static char *avail, *limit;
	
	n = (n + sizeof(char *) - 1)&~(sizeof(char *) - 1);
	if (avail + n >= limit) {
		avail = (char *)malloc(n + 4*1024);
		assert(avail);
		limit = avail + n + 4*1024;
	}
	avail += n;
	return avail - n;
}

/* append - append a node with string str onto list, return new list */	
static List append(str, list) char *str; List list; {
	List p = (List) alloc(sizeof *p);

	p->str = str;
	if (list) {
		p->link = list->link;
		list->link = p;
	} else
		p->link = p;
	return p;
}

/* basename - return base name for name, e.g. /usr/drh/foo.c => foo */
char *basename(name) char *name; {
	char *s, *b, *t = 0;

	for (b = s = name; *s; s++)
		if (*s == '/') {
			b = s + 1;
			t = 0;
		} else if (*s == '.')
			t = s;
	s = strsave(b);
	if (t)
		s[t-b] = 0;
	return s;
}

/* callsys - fork and execute the command described by argv[0...], return status */
static int callsys(argv) char **argv; {
	int n, m = 0, status = 0, pid;

	cprint(argv, 0);
	if (verbose >= 2)
		return 0;
	switch (pid = fork()) {
	case -1:
		fprintf(stderr, "%s: no more processes\n", progname);
		return 100;
	case 0:
		execute(argv);
	}
	while ((n = wait(&m)) != pid && n != -1)
		status |= m;
	status |= m;
	if (status&0377) {
		fprintf(stderr, "%s: fatal error in %s\n", progname, argv[0]);
		status |= 0400;
	}
	return (status>>8)&0377;
}

/* concat - return concatenation of strings s1 and s2 */
char *concat(s1, s2) char *s1, *s2; {
	int n = strlen(s1);
	char *s = alloc(n + strlen(s2) + 1);

	strcpy(s, s1);
	strcpy(s + n, s2);
	return s;
}

/* compile - compile src into dst, return status */
static int compile(src, dst) char *src, *dst; {
	int n, status;

	compose(com, clist, append(src, 0), append(dst, 0));
	status = callsys(av);
	if (status && *src == '-') {
		char buf[1024];
		while ((n = read(0, buf, sizeof buf)) > 0)
			;
	}
	return status;
}

/* compose - compose cmd into av substituting a, b, c for $1, $2, $3, resp. */
static void compose(cmd, a, b, c) char *cmd[]; List a, b, c; {
	int i, j;
	List lists[3];

	lists[0] = a;
	lists[1] = b;
	lists[2] = c;
	for (i = j = 0; cmd[i]; i++)
		if (cmd[i][0] == '$' && isdigit(cmd[i][1])) {
			int k = cmd[i][1] - '0';
			assert(k >= 1 && k <= 3);
			if (b = lists[k-1])
				do {
					b = b->link;
					assert(j < ac);
					av[j++] = b->str;
				} while (b != lists[k-1]);
		} else if (*cmd[i]) {
			assert(j < ac);
			av[j++] = cmd[i];
		}
	av[j] = 0;
}

/* cprint - print the command described by argv[0...] followed by str or \n */
static void cprint(argv, str) char *argv[], *str; {
	if (verbose) {
		fprintf(stderr, "%s", *argv++);
		while (*argv)
			fprintf(stderr, " %s", *argv++);
		if (str == 0)
			str = "\n";
		fprintf(stderr, str);
	}
}

/* error - issue error msg according to fmt, bump error count */
static void error(fmt, msg) char *fmt, *msg; {
	fprintf(stderr, "%s: ", progname);
	fprintf(stderr, fmt, msg);
	fprintf(stderr, "\n");
	errcnt++;
}

/* execute - replace this process by the command described by argv[0...] */
static void execute(argv) char *argv[]; {
	if (verbose >= 2)
		return;
	execv(argv[0], argv);
	fprintf(stderr, "%s: can't execute `%s'\n", progname, argv[0]);
	fflush(stdout);
	exit(100);
}

/* exists - is `name' readable? issue message if not */
static int exists(name) char *name; {
	if (verbose > 1 || access(name, 4) == 0)
		return 1;
	error("can't read `%s'", name);
	return 0;
}

/* filename - process file name argument `name', return status */
static int filename(name, base) char *name, *base; {
	int status = 0;

	if (base == 0)
		base = basename(name);
	switch (suffix(name)) {
	case 'c':
		if (!exists(name))
			break;
		compose(cpp, plist, append(name, 0), 0);
		if (Eflag) {
			status = callsys(av);
			break;
		}
		cprint(av, " | ");
		if (verbose <= 1) {
			int fd[2], pid;
			if (pipe(fd) < 0) {
				error("can't create preprocessor-compiler pipe\n", 0);
				exit(1);
			}
			switch (pid = fork()) {
			case -1:
				fprintf(stderr, "%s: no more processes\n", progname);
				return 100;
			case 0:
				close(1);
				dup(fd[1]);
				close(fd[0]);
				close(fd[1]);
				execute(av);
				assert(0);	/* no return from execute */
			}
			close(0);
			dup(fd[0]);
			close(fd[0]);
			close(fd[1]);
		}
		if (Sflag)
			status = compile("-", outfile ? outfile : concat(base, ".s"));
		else if ((status = compile("-", tempname)) == 0)
			return filename(tempname, base);
		break;
	case 'i':
		if (!exists(name) || Eflag)
			break;
		if (Sflag)
			status = compile(name, outfile ? outfile : concat(base, ".s"));
		else if ((status = compile(name, tempname)) == 0)
			return filename(tempname, base);
		break;
	case 's':
		if (Eflag)
			break;
		if (!Sflag) {
			char *ofile = cflag && outfile ? outfile : concat(base, ".o");
			compose(as, alist, append(name, 0), append(ofile, 0));
			status = callsys(av);
			if (!find(ofile, llist[1]))
				llist[1] = append(ofile, llist[1]);
		}
		break;
	case 'o':
		if (!find(name, llist[1]))
			llist[1] = append(name, llist[1]);
		break;
	case -1:
		if (Eflag) {
			compose(cpp, plist, append(name, 0), 0);
			status = callsys(av);
		} /* else fall thru */
	default:
		llist[1] = append(name, llist[1]);
		break;
	}
	if (status)
		errcnt++;
	return status;
}

/* find - find 1st occurrence of str in list, return list node or 0 */
static List find(str, list) char *str; List list; {
	List b;
	
	if (b = list)
		do {
			if (strcmp(str, b->str) == 0)
				return b;
		} while ((b = b->link) != list);
	return 0;
}

static
void
interrupt(n) {
	rm(rmlist);
	exit(n = 100);
}

/* opt - process option in arg */
static void opt(arg) char *arg; {
	switch (arg[1]) {	/* multi-character options */
	case 'W':	/* -Wxarg */
		if (arg[2] && arg[3])
			switch (arg[2]) {
			case 'o':
				if (option(&arg[3]))
					return;
				break;
			case 'p':
				plist = append(&arg[3], plist);
				return;
			case 'f':
				if (strcmp(&arg[3], "-C") || option("-b")) {
					clist = append(&arg[3], clist);
					return;
				}
				break; /* and fall thru */
			case 'a':
				alist = append(&arg[3], alist);
				return;
			case 'l':
				llist[0] = append(&arg[3], llist[0]);
				return;
			}
		fprintf(stderr, "%s: %s ignored\n", progname, arg);
		return;
	case 'd':	/* -dn */
		arg[1] = 's';
		/* fall thru */
	case 't':	/* -t -tname */
		clist = append(arg, clist);
		return;
#ifdef sun
	case 'f':	/* -f68881 */
		if (!option(arg))
			fprintf(stderr, "%s: %s ignored\n", progname, arg);
		return;
#endif
	case 'p':	/* -p -pg */
		if (option(arg))
			clist = append(arg, clist);
		else
			fprintf(stderr, "%s: %s ignored\n", progname, arg);
		return;
	case 'D':	/* -Dname -Dname=def */
	case 'U':	/* -Uname */
	case 'I':	/* -Idir */
		plist = append(arg, plist);
		return;
	case 'B':	/* -Bdir/ -Bstatic -Bdynamic */
#ifdef sun
		if (strcmp(arg, "-Bstatic") == 0 || strcmp(arg, "-Bdynamic") == 0)
			llist[1] = append(arg, llist[1]);
		else
#endif	
		{
		static char *path;
		if (path)
			error("-B overwrites earlier option", 0);
		path = arg + 2;
		com[0] = concat(path, "rcc");
		if (path[0] == 0)
			error("missing directory in -B option", 0);
		}
		return;
	}
	if (arg[2] == 0)
		switch (arg[1]) {	/* single-character options */
		case 'S':
			Sflag++;
			return;
		case 'O':
			fprintf(stderr, "%s: %s ignored\n", progname, arg);
			return;
		case 'A': case 'n': case 'w': case 'P':
			clist = append(arg, clist);
			return;
		case 'g': case 'b':
			if (option(arg))
				clist = append(arg[1] == 'g' ? "-g2" : arg, clist);
			else
				fprintf(stderr, "%s: %s ignored\n", progname, arg);
			return;
		case 'G':
			if (option(arg)) {
				clist = append("-g3", clist);
				llist[0] = append("-N", llist[0]);
			} else
				fprintf(stderr, "%s: %s ignored\n", progname, arg);
			return;
		case 'E':
			Eflag++;
			return;
		case 'c':
			cflag++;
			return;
		case 'C':
			plist = append(arg, plist);
			return;
		case 'N':
			if (strcmp(basename(cpp[0]), "gcc-cpp") == 0)
				plist = append("-nostdinc", plist);
			include[0] = 0;
			return;
		case 'v':
			if (verbose++ == 0) {
				if (strcmp(basename(cpp[0]), "gcc-cpp") == 0)
					plist = append(arg, plist);
				clist = append(arg, clist);
				fprintf(stderr, "%s version %d.%d\n", progname, VERSION>>8, VERSION&0xff);
			}
			return;
		}
	if (cflag || Sflag || Eflag)
		fprintf(stderr, "%s: %s ignored\n", progname, arg);
	else
		llist[1] = append(arg, llist[1]);
}

/* rm - remove files in list */
static void rm(list) List list; {
	if (list) {
		List b = list;
		if (verbose)
			fprintf(stderr, "rm");
		do {
			if (verbose)
				fprintf(stderr, " %s", b->str);
			unlink(b->str);
		} while ((b = b->link) != list);
		if (verbose)
			fprintf(stderr, "\n");
	}
}

/* strsave - return a saved copy of string str */
char *strsave(str) char *str; {
	return (char *)strcpy(alloc(strlen(str)+1), str);
}

/* suffix - return the 1-character suffix of name, e.g. /usr/drh/foo.c => 'c' */
int suffix(name) char *name; {
	char *t = 0, *s;

	for (s = name; *s; s++)
		if (*s == '/')
			t = 0;
		else if (*s == '.')
			t = s + 1;
	if (t && t[1] == 0)
		return t[0];
	return -1;
}
